File: Xenon/Xenon/RusherGenerator.ts
import CActor = require("Actor");
import CRusher = require("Rusher");
import gsCControls = require("Controls");
import gsCTimer = require("Timer");
import enums = require("Enums");

class CRusherGenerator extends CActor {
    
    private RUSHER_TOTAL: number = 6;	// total segments in chain
    private RUSHER_DELAY: number = 0.5;	// time delay between generation

    //-------------------------------------------------------------

    private m_rushers_created: number;
    private m_delay_timer:gsCTimer;

    constructor() {
        super();
        this.m_rushers_created = 0;
        //this.m_timer = new gsCTimer();
        this.m_delay_timer = new gsCTimer();
        this.m_name = "RusherGenerator";
    }

    public getActorInfo()
    {
        this.m_actorInfo = this.m_scene.GetlistOfActors();
        return this.m_actorInfo.GetActorInfoListItem(enums.ActorInfoType.INFO_RUSHER_GENERATOR);
    }


    public activate(): boolean {
        if (!this.isActive()) {
            this.m_timer.start();
            this.m_delay_timer.start();
        }

        return super.activate();
    }

    public update(controls: gsCControls, gameTime: gsCTimer) {
        this.gameTime = gameTime;
        if (this.m_delay_timer.getTime() < this.RUSHER_DELAY)
        {
            return true;
        }
        this.m_delay_timer.start();

        var r: CRusher = new CRusher();
        this.m_scene.addActor(r);
        r.setPosition(this.getPosition());
        r.setVelocity(this.getVelocity());
        r.activate();

        this.m_rushers_created++;
        if (this.m_rushers_created >= this.RUSHER_TOTAL) {
            super.kill();
        }
        return true;
    }
}

export = CRusherGenerator;

File: Xenon-Original_C++_Code/xenon/source/rushergenerator.cpp
//-------------------------------------------------------------
//
// Class:	CRusherGenerator
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CActor
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

CRusherGenerator::CRusherGenerator()
{
	m_rushers_created = 0;
}

//-------------------------------------------------------------

CRusherGenerator::~CRusherGenerator()
{
}

//-------------------------------------------------------------

bool CRusherGenerator::activate()
{
	if (!isActive()) {
		m_timer.start();
		m_delay_timer.start();
		}

	return CActor::activate();
}

//-------------------------------------------------------------

bool CRusherGenerator::update(Controls *controls)
{
	if (m_delay_timer.getTime() < RUSHER_DELAY)
		return true;

	m_delay_timer.start();
	
	CRusher *r = new CRusher;
	m_scene->addActor(r);
	r->setPosition(getPosition());
	r->setVelocity(getVelocity());
	r->activate();

	m_rushers_created++;
	if (m_rushers_created >= RUSHER_TOTAL)
		kill();

	return true;
}

//-------------------------------------------------------------
